gnucash/report/trep-engine.scm

```diff
@@ -154,7 +162,51 @@ in the Options panel."))
         (cons #\f (G_ "Frozen"))
         (cons #\v (G_ "Voided"))))
 
-(define (sortkey-list split-action?)
+;; Hashtable to memorize both the sortvalue and rendering of the
+;; split's matching hashtag 
+(define hashtag-htable (make-hash-table))
+
+;; Create a function which helps find a split's hashtag as per
+;; various user-defined options in pagename-hashtag section.
+;; Memorize both the sortvalue and rendering in hash table
+;; for later retrieval because get-hashtag-from-split is slow. 
+;; Returns either the sort value or rendered value as per sortvalue?
+(define (split-hashtag split options sortvalue?)
+    (or (let ((tag (hash-ref hashtag-htable (gncSplitGetGUID split))))
+          (and tag (if sortvalue? (car tag) (cadr tag))))
+        (let ((tag (get-hashtag-from-split split options)))
+          (hash-set! hashtag-htable (gncSplitGetGUID split) tag)
+          (if sortvalue? (car tag) (cadr tag)))))
+
+;; Finds a split's hashtag as per user-defined options.
+;; Returns a list with both (hashtag-sortvalue hashtag-rendered)
+(define (get-hashtag-from-split split options)
+  (define (opt-val section name)
+    (gnc-optiondb-lookup-value (gnc:optiondb options) section name))
+  (let* ((hashtag-prefix (opt-val pagename-hashtag optname-hashtag-prefix))
+         (remove-htp? (opt-val pagename-hashtag optname-remove-htp))
+         (no-match-heading (opt-val pagename-hashtag optname-no-match-heading))
+         (append-htp? (opt-val pagename-hashtag optname-append-htp))
+         (htp (if (string-null? (string-trim hashtag-prefix))
+                   "#" (string-trim hashtag-prefix)))
+         (htp-sanitized (regexp-substitute/global #f "[#-.]|[[-^]|[?|{}]" htp
+                          'pre (lambda (m) (string-append "\\" (match:substring m))) 'post))
+         (regexp (make-regexp (string-append htp-sanitized "[^ ]*")))
```

Comment:
IIUC `make-regexp` is slow... it should be generated once, and reused within a trep-renderer session